src/lib/common/include/sol-update.h

```diff
@@ -0,0 +1,185 @@
+ *
+ * @return handle of this update task, or NULL if couldn't start the task.
+ * Note that handle will be invalid after @c cb is called, or if it was cancelled.
+ */
+struct sol_update_handle *sol_update_check(const char *url,
+    void (*cb)(void *data, int status, const struct sol_update_info *response),
+    const void *data);
+
+/**
+ * Fetch update file using information given via @c sol_update_info
+ *
+ * Fetch will get the update file and check its hash against the one informed
+ * via @c sol_update_info. If everything is OK, callback @c cb will be
+ * called with downloaded file path.
+ *
+ * @param info sol_update_info containing information about update file.
+ * Usually is obtained via @c sol_update_check call.
+ * @param cb callback called to inform fetch completion or failure. If
+ * status < 0, then something went wrong and @c file_path is undefined.
+ * If status == 0, fetch went OK and @c file_path contains update_file path.
+ * @param data user specified data to be sent to callback @c cb
+ * @param resume NOT IMPLEMENTED. If false, any previous data downloaded
+ * shall be discarded. Otherwise, will resume a previous cancelled fetch.
+ *
+ * @return handle of this update task, or NULL if couldn't start the task.
+ * Note that handle will be invalid after @c cb is called, or if it was cancelled.
+ */
+/* TODO fetch should also verify a file signature, to be sure that we
+ * downloaded the file from a trusted source */
+struct sol_update_handle *sol_update_fetch(const struct sol_update_info *info,
+    void (*cb)(void *data, int status, const char *file_path),
+    const void *data, bool resume);
+
+/**
+ * Install an update file.
+ *
+ * Install process vary depending on update module being used, but
+ * all shall give feedback of success or not on @c cb callback.
+ *
+ * @param file_path path of update file. Usually should be the one got via
+ * @c sol_update_fetch call.
+ * @param cb callback called to inform installation success or failure.
+ * If status < 0, then something went wrong. If status == 0, installation
+ * completed successfuly.
+ *
+ * @return handle of this update task, or NULL if couldn't start the task.
+ * Note that handle will be invalid after @c cb is called, or if it was cancelled.
+ */
+struct sol_update_handle *sol_update_install(const char *file_path,
+    void (*cb)(void *data, int status),
+    void *data);
+
+/**
+ * Cancel an ongoing check, fetch or install task.
+ * Note that cancelling the task may not be possible at any moment, depending
+ * on update module being used.
+ *
+ * @param handle handle of update task to be cancelled.
+ *
+ * @return true if current task was successfuly cancelled.
+ *
+ * @note If cancel was successful, then @c handle becomes invalid. If not,
+ * task callback shall still be called.
+ */
+/* TODO note above may yeld to memory leak if app is closed before
+ * completing the task. Shall we worry about that? */
```

Comment:
not clear what you mean, you should be able to cancel it. If the backend can't for some reason, then it should use some other trick such as "atexit()"